cudax/include/cuda/experimental/__memory_resource/cuda_memory_pool.h

```diff
@@ -0,0 +1,265 @@
+  {
+    _LIBCUDACXX_ASSERT(_CUDA_VMR::__device_supports_pools(__device_id), "cudaMallocAsync not supported");
+    _LIBCUDACXX_ASSERT(__cuda_supports_export_handle_type(__device_id, __properties.allocation_handle_type),
+                       "Requested IPC memory handle type not supported");
+
+    ::cudaMemPoolProps __pool_properties{};
+    __pool_properties.allocType     = ::cudaMemAllocationTypePinned;
+    __pool_properties.handleTypes   = ::cudaMemAllocationHandleType(__properties.allocation_handle_type);
+    __pool_properties.location.type = ::cudaMemLocationTypeDevice;
+    __pool_properties.location.id   = __device_id;
+    ::cudaMemPool_t __cuda_pool_handle{};
+    _CCCL_TRY_CUDA_API(::cudaMemPoolCreate, "Failed to call cudaMemPoolCreate", &__cuda_pool_handle, &__pool_properties);
+
+    // CUDA drivers before 11.5 have known incompatibilities with the async allocator.
+    // We'll disable `cudaMemPoolReuseAllowOpportunistic` if cuda driver < 11.5.
+    // See https://github.com/NVIDIA/spark-rapids/issues/4710.
+    int __driver_version = 0;
+    _CCCL_TRY_CUDA_API(::cudaDriverGetVersion, "Failed to call cudaDriverGetVersion", &__driver_version);
+
+    constexpr int __min_async_version = 11050;
+    if (__driver_version < __min_async_version)
+    {
+      int __disable_reuse = 0;
+      _CCCL_TRY_CUDA_API(
+        ::cudaMemPoolSetAttribute,
+        "Failed to call cudaMemPoolGetAttribute",
+        __cuda_pool_handle,
+        ::cudaMemPoolReuseAllowOpportunistic,
+        &__disable_reuse);
+    }
+
+    _CCCL_TRY_CUDA_API(
+      ::cudaMemPoolSetAttribute,
+      "Failed to call cudaMemPoolReuseAllowOpportunistic",
+      __cuda_pool_handle,
+      ::cudaMemPoolAttrReleaseThreshold,
+      &__properties.release_threshold);
+
+    // allocate the requested initial size to pprime the pool
+    if (__properties.initial_pool_size != 0)
+    {
+      void* __ptr{nullptr};
+      _CCCL_TRY_CUDA_API(
+        ::cudaMallocAsync,
+        "cuda_memory_pool failed allocate the initial pool size",
+        &__ptr,
+        __properties.initial_pool_size,
+        ::cudaStream_t{0});
+
+      _CCCL_ASSERT_CUDA_API(
+        ::cudaFreeAsync, "cuda_memory_pool failed to free the initial pool allocation", __ptr, ::cudaStream_t{0});
+    }
+    return __cuda_pool_handle;
+  }
+
+public:
+  //! @brief Constructs a \c cuda_memory_pool with the optionally specified initial pool size and release threshold.
+  //! If the pool size grows beyond the release threshold, unused memory held by the pool will be released at the next
+  //! synchronization event.
+  //! @throws cuda_error if the CUDA version does not support ``cudaMallocAsync``
+  //! @param __device_id The device id of the device the stream pool is constructed on.
+  //! @param __pool_properties Optional, additional properties of the pool to be created
+  explicit cuda_memory_pool(const int __device_id, cuda_memory_pool_properties __properties = {})
```

Comment:
Do we want this to be able to take a naked device ordinal, or would device_ref be more appropriate here?